cache/cache_reservation_manager.h

```diff
@@ -172,20 +208,59 @@ class CacheReservationManager
   CacheKey cache_key_;
 };
 
-// CacheReservationHandle is for managing the lifetime of a cache reservation
-// This class is NOT thread-safe
-template <CacheEntryRole R>
-class CacheReservationHandle {
+class ConcurrentCacheReservationManager : public CacheReservationManager {
  public:
-  // REQUIRES: cache_res_mgr != nullptr
-  explicit CacheReservationHandle(
+  explicit ConcurrentCacheReservationManager(
+      std::shared_ptr<CacheReservationManager> cache_res_mgr) {
+    cache_res_mgr_ = cache_res_mgr;
+  }
+  ConcurrentCacheReservationManager(const ConcurrentCacheReservationManager &) =
+      delete;
+  ConcurrentCacheReservationManager &operator=(
+      const ConcurrentCacheReservationManager &) = delete;
+  ConcurrentCacheReservationManager(ConcurrentCacheReservationManager &&) =
+      delete;
+  ConcurrentCacheReservationManager &operator=(
+      ConcurrentCacheReservationManager &&) = delete;
+
+  ~ConcurrentCacheReservationManager() override {}
+
+  inline Status UpdateCacheReservation(std::size_t new_memory_used) override {
+    std::lock_guard<std::mutex> lock(cache_res_mgr_mu_);
+    return cache_res_mgr_->UpdateCacheReservation(new_memory_used);
+  }
+  inline Status MakeCacheReservation(
       std::size_t incremental_memory_used,
-      std::shared_ptr<CacheReservationManager> cache_res_mgr);
+      std::unique_ptr<CacheReservationManager::CacheReservationHandle> *handle)
+      override {
+    std::lock_guard<std::mutex> lock(cache_res_mgr_mu_);
+    Status s = cache_res_mgr_->UpdateCacheReservation(
+        cache_res_mgr_->GetTotalMemoryUsed() + incremental_memory_used);
+    (*handle).reset(new CacheReservationHandle(incremental_memory_used,
+                                               this->shared_from_this()));
 
-  ~CacheReservationHandle();
+    return s;
```

Comment:
[Note]
 The reason why I didn't do 
```
std::lock_guard<std::mutex> lock(cache_res_mgr_mu_);
return cache_res_mgr_->MakeCacheReservation(..);
```

is that the handle needs to point to a ConcurrentCacheReservationManager for synchronized ReleaseCacheReservation on handle's destruction. It feels like a "leak" of opportunity of data race into CRMHandle class out of CRM for the benefit of RAII and now we are paying for that by this code duplication treatment.

Reply:
Got you - will try out your suggestion.

Reply:
Okay - now I can see how bad Github UI is. "Got you - will try out your suggestion." was meant to be reply to other threads. And I need to look into your "see below" first.

Reply:
Okay - will still try your suggestion after reading!

Reply:
Good point on making ConcurrentCacheReservationManager::CacheReseravationHandle a decorator of CacheReservationManager::CacheReseravationHandle, which climates two design weirdness I previously pondered. Done.